README.md
[![Code QC](https://github.com/milivojevicu/mpcover/actions/workflows/check.yml/badge.svg)](https://github.com/milivojevicu/mpcover/actions/workflows/check.yml)

# MPCover: Album cover display for [MPD](https://github.com/MusicPlayerDaemon/MPD)

Python program for displaying album covers of music currently playing through MPD using tkinter.

<p align="center">
  <img src="images/screenshot_0.png" width="500"/>
</p>

## Installation

Install the package from PyPI using `pip`:

```bash
pip install mpcover
```

## Configuration

The configuration file should be located in the user home directory with the name ".mpcover.init".

For more information on where the user home directory is,
reference [`os.path.expanduser`](https://docs.python.org/3/library/os.path.html#os.path.expanduser).

Example configuration file:

```ini
[connection]
# Connection settings. The password is optional, to leave it unset simply remove
# the "password = ..." line from the configuration file.
host = localhost
port = 6600
password = example_password

[logging]
level = info

[style]
background = #141414
padding = 12

[binds]
# The values should be `tkinter` key bind strings.
refresh = r
quit = q

[other]
image_size = 512
```



CHANGELOG.md
## Version 0.3.0

- [x] Configurable style.
- [x] Configurable image size.
- [x] Quit keybind.

## Version 0.2.0

- [x] Handle no album art being available when the file is missing.
- [x] Keybinds for forcing album art refresh.
- [x] Add ability to change settings through a config file.

## Version 0.1.0

- [x] Implement main functionality:
	- [x] Create window with `tkinter`,
	- [x] Get album art from MPD,
	- [x] Display album art,
	- [x] Use MPDs `idle` command to detect player and playlist changes and get new album art.
- [x] Add support for password authentication.
- [x] Add ability to change settings through command line arguments.
- [x] Handle no album art being available when no song is playing.


mpcover/__main__.py
#!python

import argparse
import configparser
import logging

from .config import get_config
from .gui import init

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)


def parse_arguments(config: configparser.ConfigParser) -> argparse.Namespace:
    """
    Parse arguments from the command line.

    :arg config: Configuration object read from a file. Used to set default values.
        Values from the command line override the values from the config file.

    :return: `argparse.Namespace` object with the parsed arguments.
    """

    parser = argparse.ArgumentParser(
        description="Album cover viewer for MPD.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )

    parser.add_argument(
        "-a",
        "--address",
        metavar="ADDRESS",
        type=str,
        default=config.get("connection", "host"),
        help="MPD server IP address",
    )

    parser.add_argument(
        "-p",
        "--port",
        metavar="PORT",
        type=int,
        default=config.get("connection", "port"),
        help="MPD server port",
    )

    parser.add_argument(
        "-s",
        "--pass",
        metavar="PASSWORD",
        dest="password",
        type=str,
        default=config.get("connection", "password", fallback=None),
        help="password for auth with the MPD server",
    )

    return parser.parse_args()


def run():
    """
    Entry point. Called when the `mpcover` command is called.
    """

    config = get_config()
    arguments = parse_arguments(config)
    address = arguments.address, arguments.port
    init(address, arguments.password)


if __name__ == "__main__":
    run()


mpcover/gui/root.py
import configparser
import io
import tkinter as tk
from logging import getLogger
from multiprocessing import Process, Queue
from typing import Any, Dict, Optional, Tuple

from PIL import Image, ImageTk

from ..config import get_config
from ..connection import Connection
from ..controler import Controler

logger = getLogger(__name__)


class Root(tk.Tk):
    """
    Main graphical user inteface class.

    :arg address: A string IP address and an integer port where MPD is running.
    :arg password: Password for auth with the MPD server.

    :var connection: Connection to MPD.
    :var controler: Interface to MPD.

    :var canvas: Cavnas on which the album art is drawn.
    :var canvas_width: Width of the canvas. Updated with window resize events.
    :var canvas_height: Height of the canvas. Updated with window resize events.

    :var album_art_original: Album art originally downloaded from MPD, resized
        to 512x512 if larger then that.
    :var album_art: Album art displayed on the canvas.
    :var album_queue: Queue used for communication between the main GUI process
        and the player change monitoring process.
    :var album: Name of the currently playing album.
    :var album_connection: Connection to MPD used for monitoring player changes.
    :var album_controler: Interface to MPD used for monitoring player changes.
    :var album_process: Process that waits for player changes using the `idle`
        command.
    """

    def __init__(self, address: Tuple[str, int], password: Optional[str]):
        super().__init__()

        config: configparser.ConfigParser = get_config()

        # Connect to MPD.
        self.__connection: Connection = Connection(*address)
        self.__controler: Controler = Controler(self.__connection, password)

        # Read configuration.
        self.__color_background: str = config.get("style", "background")
        self.__padding: int = config.getint("style", "padding")
        self.__image_size: int = config.getint("other", "image_size")

        # Configure window.
        self.title("MPCover")
        self.configure(background=self.__color_background)
        # The geometry doesn't always apply. There is a lower limit for window size so
        # if the image size is _really_ small, the window won't resize to it perfectly.
        self.geometry(
            str(self.__image_size + 2 * self.__padding)
            + "x"
            + str(self.__image_size + 2 * self.__padding)
        )

        # Window close hook.
        self.protocol("WM_DELETE_WINDOW", self.close)

        # Canvas for album art.
        self.__canvas: tk.Canvas = tk.Canvas(
            self, highlightthickness=0, background=self.__color_background
        )
        self.__canvas.grid(column=0, row=0, padx=self.__padding, pady=self.__padding, sticky="nwes")
        self.__canvas.bind("<Configure>", self.__display_album_art)

        # Canvas size, updated on `<Configure>` events.
        self.__canvas_width: int = 100
        self.__canvas_height: int = 100

        # Max wights on canvas at (0, 0).
        self.columnconfigure(0, weight=1)
        self.rowconfigure(0, weight=1)

        # Original image retrieved from MPD. Kept in order to stop repeated
        # resizing from lowering the quality of the image by always using
        # the original image when resizing instead of the displayed one.
        # Always first resized for better performance with rapid window size changes.
        self.__album_art_original: Optional[Image.Image] = None
        # Image rescaled for window size.
        self.__album_art: Optional[ImageTk.PhotoImage] = None
        # Queue for passing changes from the album process.
        self.__album_queue: Queue = Queue()
        # Album name. Used for tracking when the album changes to trigger
        # downloading new album art.
        self.__album: str = ""

        # Player change detection. Starts a new process for idling while
        # waiting for a player change to occur. This process fills a queue
        # with detected changes while the `idle_player_change` constantly
        # reads and empties the queue while initiating the album art updates.
        # Check for updates when idle.
        self.after(50, self.idle_player_change)
        # New connection for idling.
        self.__album_connection: Connection = Connection(*address)
        self.__album_controler: Controler = Controler(self.__album_connection, password)
        # Process to idle independently of the main GUI process.
        self.__album_process: Process = Process(
            target=self.__album_controler.idle,
            args=(self.__album_queue, "player", "playlist"),
        )
        self.__album_process.start()

        # Initial album art get.
        self.__get_album_art()

        # Handle whole window keybinds.
        # The lambda functions are there just to capture the `event` argument without passing
        # it to the actaul functions since they don't need it.
        self.bind(config.get("binds", "refresh"), lambda event: self.__get_album_art())
        self.bind(config.get("binds", "quit"), lambda event: self.close())

    def close(self):
        """
        Called by `tkinter` on window close. The `tkinter` window can close on
        its own, this only overrides the closing process to terminate the
        player change monitoring process before exiting.
        """

        self.__album_process.terminate()
        self.destroy()

    def idle_player_change(self):
        """
        Called by the album process when a change in the player is detected
        (seek, change song, etc.)
        """

        # Read queue if it's not empty and update album art.
        if not self.__album_queue.empty():
            self.__album_queue.get()
            self.__get_album_art()

        # Run again in .5 seconds.
        self.after(50, self.idle_player_change)

    def __get_album_art(self):
        """
        Downloads album art for the current song if the album has changed
        since the last call. Resizes the downloaded art if it's larger then that.
        Calls `display_album_art`.
        """

        # Don't display album art if the current song is stop.
        playback_state: str = self.__controler.status()["state"]
        if playback_state == "stop":
            logger.debug("Song is stopped, won't display album art.")
            self.__clear_album_art()
            return

        # Get album of the currently playing song.
        current_song_data: Dict[str, Any] = self.__controler.currentsong()
        if "Album" not in current_song_data:
            # Song does not have an Album tag, just give up... for now?
            logger.debug("Current song does not have an album tag, giving up...")
            self.__clear_album_art()
            return
        album: str = current_song_data["Album"]

        # Return if it's the same album as the currently loaded album art.
        if album == self.__album:
            return
        self.__album = album

        logger.debug("Getting new album art.")

        # Get album art from MPD.
        data: Optional[bytes] = self.__controler.albumart()
        if data is not None:
            # Open as Pillow image.
            self.__album_art_original = Image.open(io.BytesIO(data))
            logger.debug(type(self.__album_art_original))
            # Resize if image is large.
            if self.__album_art_original.size > (self.__image_size, self.__image_size):
                self.__album_art_original = self.__album_art_original.resize(
                    (self.__image_size, self.__image_size)
                )
        else:
            self.__album_art_original = None

        # Display the new album art.
        self.__display_album_art()

    def __display_album_art(self, event: Optional[tk.Event] = None):
        """
        Displays album art. Called by `get_album_art` and by `tkinter` on
        window resize.

        :arg event: Info about the window resize event.
        """

        # Get canvas size if it was resized.
        if event is not None:
            self.__canvas_width = event.width
            self.__canvas_height = event.height

        self.__clear_album_art()

        # If no album art is available, leave the canvas blank.
        if self.__album_art_original is None:
            return

        # Get square dimentions (assuming that album art is square).
        size = min(self.__canvas_width, self.__canvas_height)

        # Convert the album art image to a Tk Photo Image and resize
        # to match canvas size.
        self.__album_art = ImageTk.PhotoImage(self.__album_art_original.resize((size, size)))

        # Draw album art to canvas.
        self.__canvas.create_image(
            self.__canvas_width // 2,
            self.__canvas_height // 2,
            image=self.__album_art,
        )

    def __clear_album_art(self):
        self.__album = ""
        self.__canvas.delete("all")


mpcover/controler.py
from __future__ import annotations

import logging
import re
import sys
from multiprocessing import Queue
from typing import Callable, Dict, Iterable, List, Optional, Tuple, Union

from .connection import Connection

logger = logging.getLogger(__name__)


def generic_command(method: Callable) -> Callable:
    """
    Decorator for generic command methods. Uses the method name as the
        command.

    :arg method: The method being decorated.

    :return: Lambda function for running the generic command and parsing its
        output.
    """

    return lambda self: self.parse_items(list(self.run(method.__code__.co_name)))


class Controler:
    """
    MPD controler class. Containes methods that send commands and return
    parsed responses.

    :var connection: Connection to the MPD server.
    :var mpd_version: Version of the MPD server, recieved on connection open
        from the server.
    """

    # Item regex. Group 1 is the key and group 2 is the value.
    RE_ITEM = re.compile("^([A-Za-z-_]*): (.*)$")
    # Regex patterns for primitive data types. Used when parsing response items.
    RE_INTEGER = re.compile("^[0-9]+$")
    RE_FLOATING = re.compile("^[0-9.]+$")

    def __init__(self, connection: Connection, password: Optional[str] = None):
        self.__connection = connection
        self.__password = password

        self.__mpd_version = self.__connection.recv()[7:-1].decode()
        logging.info("MPD %s", self.__mpd_version)
        self.__auth()

    def __del__(self):
        """Destructor, closes connection to MPD server."""

        self.__connection.close()

    def __auth(self):
        """
        Perform authentication.

        :arg password: Optional password.
        """

        if self.__password is not None:
            self.__connection.send(f'password "{self.__password}"'.encode())
            response = self.__connection.recv().decode()
            if response != "OK\n":
                logger.critical("Failed to authenticate with message: %s", response[:-1])
                logger.critical("Exiting...")
                sys.exit(201)
            else:
                logger.info("Successful authentication.")
        else:
            logger.info("No password provided, assuming successful connection.")

    def run(self, command: str, *args: str) -> Iterable[bytes]:
        """
        Encode and send a command to the MPD server, then deocde and yeild
        the response.

        :arg command: Command string.
        :arg *args: Positional arguments for the command.

        :return: Yeilds bytes. Finished when success or error item is
            encounterd.
        """

        # Encode command string.
        data = command.encode()

        # Iterate of arguments.
        for arg in args:
            # If an argument is not a string, convert it to a string.
            if not isinstance(arg, str):
                arg = str(arg)

            # If an argument contains space characters, put it in quotes.
            if " " in arg:
                arg = '"' + arg + '"'

            # Encode and append argument to command.
            data += b" " + arg.encode()

        attempts = 0

        # Set default value of respnse to an empty string,
        # equates it with getting no respnose from `recv`
        # in case of broken pipe.
        response = "".encode("UTF-8")
        while attempts < 3:
            try:
                # Send command to MPD.
                self.__connection.send(data)
                # Gather response.
                response = self.__connection.recv()
            except BrokenPipeError as error:
                logger.error(error)
            except OSError as error:
                logger.error(error)

            # No respnse, try again.
            if len(response) == 0:
                logger.debug("Got no response, attempting to reconnect.")
                attempts += 1
                # Reconnect.
                self.__connection = Connection(self.__connection.host, self.__connection.port)
                # Flush MPD version output.
                self.__connection.recv()
                # Authenticate.
                self.__auth()
                continue

            break

        yield from self.__parse_response(response)

    def __parse_item(self, item: bytes) -> Tuple[str, Union[str, int, float, bytes]]:
        """
        Generic parser for single items from a response.

        :return: A tuple of key and value.
        """

        # Handle binary items.
        if item[:13] == b"binary_data: ":
            return "binary_data", item[13:]

        # Get groups.
        match = self.RE_ITEM.match(item.decode())

        if match is None:
            return "", ""

        # Evaluate value.
        value = match.group(2)
        if self.RE_INTEGER.match(value):
            value = int(value)
        elif self.RE_FLOATING.match(value):
            value = float(value)

        return match.group(1), value

    def parse_items(self, items: List[bytes]) -> Dict[str, Union[str, int, float, bytes]]:
        """
        Generic response item parser.

        :return: A dictionary with parsed items.
        """

        return dict([self.__parse_item(item) for item in items])

    def __parse_response(self, response: bytes) -> Iterable[bytes]:
        """
        Processs the response sent by MPD. Yields items.

        :arg response: Response from MPD. Can be empty in case of connection issues.

        :return: Yeilds byte arrays, one per item in response. No response from MPD returns an
            empty list.
        """

        # Items are spearated by '\n' characters.
        # Last item is either 'OK' on success or 'ACK ...' on error.
        # Iterate over response data.

        item = b""  # Curently processed item.
        size = -1  # Size used when reading binary items.

        for i in range(len(response)):
            # Get byte from response. Can't use for each loop because it
            # returns integers instead of bytes. Indexing without a range
            # also returns integers instead of bytes.
            byte = response[i : i + 1]

            # End of an item.
            if size < 0 and byte == b"\n":
                # Last item, command success.
                if item == b"OK":
                    return

                # Last item, command error.
                if item[:3] == b"ACK":
                    if b"you don't have permission for" in item:
                        # No password provided but auth required.
                        logger.critical(item[3:].decode())
                        sys.exit(202)
                    else:
                        # Log error and return.
                        logger.error(item[3:].decode())
                        return

                # Handle binary items.
                if item[:8] == b"binary: ":
                    yield item
                    size = int(item[8:].decode()) - 1
                    item = b"binary_data: "
                    continue

                # Not an end response item, yield.
                yield item

                # Reset item buffer and continue.
                item = b""
                continue

            # Append byte to buffer.
            item += byte
            size -= 1

        # Removed when fixing issue #6 in pull request #53. Missing album art
        # causes MPD to not return anything, so this is a workaround. The accompanying
        # change to the `run` method cathes the `OSErrror`s produced by the socket
        # and in that case this method gets no repsonse, and an empty response is passed on.
        """
        logger.critical(
            "Should not be possible to get to this point with a valid"
            " response from the server, exiting..."
        )
        sys.exit(203)
        """

        return []

    @generic_command
    def stats(self):
        """
        Get statistics.

        :return: Dictionary with statistics (`Dict[str, int]`).
        """

    @generic_command
    def status(self):
        """
        Get player and volume status.

        :return: Dictionary with status (`Dict[str, Union[str, int]]`).
        """

    @generic_command
    def currentsong(self):
        """
        Get current song info.

        :return: Dictionary with information about the currently active song (`Dict[str, Union[str, int]]`).
        """

    def albumart(self, path: Optional[str] = None) -> Optional[bytes]:
        """
        Get album art. MPD looks for a `cover.[png|jpg|tiff|bmp]` file.

        :arg path: Path to an audio file or album directory. If none is
            specified, album art for the current song will be returned.

        :return: Album art as bytes or `None` if album art does not exist.
        """

        # Get current song path if no path was provided.
        real_path: str
        if path is None:
            real_path = self.currentsong()["file"]
        else:
            real_path = path

        # Initliaize offset counter, total size and result buffer.
        offset: int = 0
        size: int = 1
        result: bytes = b""

        # Load chunks until offset reaches end of file.
        while offset < size:
            # Run command with offset.
            items = list(self.run("albumart", real_path, str(offset)))
            # If length of items is 0, an error occured. Requested album art
            # probably does not exist.
            if len(items) == 0:
                return None

            # Parse items.
            items = self.parse_items(items)
            # Get data from parsed items.
            items_binary: int = items["binary"]  # type: ignore
            items_binary_data: bytes = items["binary_data"]  # type: ignore
            items_size: int = items["size"]  # type: ignore

            # Increase offset with size of recieved binary data.
            offset += items_binary
            # Update size again since it's probably faster then an if
            # statement and MPD always returns the same size.
            size = items_size
            # Append new binary data to result.
            result += items_binary_data

        return result

    def idle(self, queue: Queue, *subsystems: str):
        """
        Removes the connection timeout and runs an `idle` command to monitor
        specified subsystems. Put the detected change on the provided queue.
        Blocking, never exits.
        """

        self.__connection.timeout = None
        while True:
            result = self.parse_items(list(self.run("idle", *subsystems)))
            logger.debug("Subsystem change detected, puting on queue.")
            queue.put(result["changed"])


mpcover/connection.py
import logging
import socket
import sys
from typing import List, Optional, Tuple

logger = logging.getLogger(__name__)


class Connection:
    """
    Connection to an MPD server.

    :arg host: String IP address.
    :arg port: Integer port.

    :var sock: Python socket to MPD using `host` and `port`.
    """

    CHUNK_SIZE = 1024

    def __init__(self, host: str, port: int):
        # Save input data.
        self.host = host
        self.port = port

        # Initialize a variable for the socket.
        self.__sock: socket.socket

        # Get info about the provided address/port.
        address_info: List[Tuple] = socket.getaddrinfo(
            host,
            port,
            socket.AF_UNSPEC,
            socket.SOCK_STREAM,
            socket.IPPROTO_TCP,
        )

        # Try different combinations until connection is established.
        for address_family, socket_kind, protocol, _, address in address_info:
            sock: Optional[socket.socket] = None
            try:
                logger.debug(
                    "Attempting to connect: %r %r %d %r.",
                    address_family,
                    socket_kind,
                    protocol,
                    address,
                )

                # Create a socket with data from `getaddrinfo`.
                sock = socket.socket(address_family, socket_kind, protocol)
                sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
                sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
                sock.settimeout(1)
                sock.connect(address)
                self.__sock = sock
            except socket.error as error:
                # Failed to connect.
                logger.debug("Attempt failed: %r.", error)
                if sock is not None:
                    sock.close()
            finally:
                # If unable to establish connection, exit.
                if self.__sock is None:
                    logger.critical("Failed to connect to %s:%d.", host, port)
                    sys.exit(101)

        # If no info was found about the address/port, exit.
        if self.__sock is None:
            logger.critical("No info found for %s:%d, can't connect.", host, port)
            sys.exit(102)

        logger.debug("Connected.")

    def __enter__(self):
        """
        Standard Python method implemented for use with the `with` statement.
        """

        return self

    def __exit__(self, exec_type, exec_value, traceback) -> bool:
        """
        Standard Python method implemented for use with the `with` statement.

        :arg exec_type: Exception type.
        :arg exec_type: Exception value.
        :arg traceback: Traceback.
        """

        logger.debug("Connection exit: %r %r %r.", exec_type, exec_value, traceback)
        self.close()

        return False

    @property
    def timeout(self) -> Optional[float]:
        return self.__sock.gettimeout()

    @timeout.setter
    def timeout(self, timeout: Optional[int]):
        self.__sock.settimeout(timeout)

    def close(self):
        """
        Close the socket/connection.
        """

        self.__sock.close()

    def send(self, send_data: bytes):
        """
        Send bytes to MPD.

        :arg send_data: Bytes to send.
        """

        send_data = send_data + "\n".encode()
        self.__sock.sendall(send_data)

    def recv(self) -> bytes:
        """
        Recieve bytes from MPD. Is blocking, but timeout is set to 1 second by
        default.

        :return: Recieved bytes.
        """

        recv_data = "".encode()

        # Recieve data in chunks.
        while True:
            # Get chunk.
            recv_datum = self.__sock.recv(self.CHUNK_SIZE)
            # Append chunk.
            recv_data += recv_datum

            # Don't exit until all data has been recieved.
            # Match for version output on connection established.
            if (
                recv_data.endswith(b"OK\n")
                or recv_data.startswith(b"OK MPD ")
                or len(recv_data) == 0
            ):
                break

        return recv_data


mpcover/config.py
import configparser
import os.path

__DEFAULTS_CONNECTION = {
    "host": "localhost",
    "port": 6600,
}

__DEFAULTS_LOGGING = {
    "level": "info",
}

__DEFAULTS_STYLE = {
    "padding": 12,
    "background": "#141414",
}

__DEFAULTS_BINDS = {
    "refresh": "r",
    "quit": "q",
}

__DEFAULTS_OTHER = {
    "image_size": "512",
}

__CONFIG = None


def get_config():
    """
    Read configuration from a file.
    """

    global __CONFIG

    if __CONFIG is not None:
        return __CONFIG

    config = configparser.ConfigParser()

    # Load default settings.
    config.read_dict({"connection": __DEFAULTS_CONNECTION})
    config.read_dict({"logging": __DEFAULTS_LOGGING})
    config.read_dict({"style": __DEFAULTS_STYLE})
    config.read_dict({"binds": __DEFAULTS_BINDS})
    config.read_dict({"other": __DEFAULTS_OTHER})

    # Read user settings from a file.
    config.read(os.path.expanduser(os.path.join("~", ".mpcover.ini")))

    __CONFIG = config

    return config


__all__ = "get_config"
